manual/manual/cmds/instrumented-runtime.etex

```diff
@@ -0,0 +1,148 @@
+can generate {\em trace files} that can then be read
+and analyzed by users in order to understand specific runtime behaviors.
+
+{\em Common Trace Format} is the general purpose binary
+format used to encode traces.
+A complete trace is comprised of:
+\begin{itemize}
+\item a {\em metadata file}, part of the OCaml distribution
+\item and a {\em trace file}, generated by the runtime\
+  in the program being traced.
+\end{itemize}
+
+For more information on {\em Common Trace Format}, see the website at
+\ifouthtml
+\ahref{https://diamon.org/ctf/}{https://diamon.org/ctf/}.
+\else
+{\tt https://diamon.org/ctf/}
+\fi
+
+\section{s:instr-runtime-enabling}{Enabling runtime instrumentation}
+
+Compiling a program and linking it with the instrumented runtime,
+can be done with {\tt -runtime-variant} flag {\tt i}:
+
+\begin{verbatim}
+       ocamlopt -runtime-variant i program.ml -o program
+\end{verbatim}
+
+The resulting program can then be traced by running it with the environment
+variable {\tt OCAML\_EVENTLOG\_ENABLED}:
+
+\begin{verbatim}
+        OCAML_EVENTLOG_ENABLED=1 ./program
+\end{verbatim}
+
+The program will execute and generate a trace file in the
+current directory.
+
+\subsection{ss:instr-runtime-prefix}{Trace filename}
+
+The default trace filename is {\tt caml-\{pid\}-eventlog}, where {\tt \{pid\}}
+is the process identifier of the traced program.
+
+This filename can also be specified using the
+{\tt OCAML\_EVENTLOG\_FILE} environment variable.
+The given path will be suffixed with {\tt .pid.eventlog}.
+
+\begin{verbatim}
+        OCAML_EVENTLOG_FILE=/tmp/some/prefix OCAML_EVENTLOG_ENABLED=1 ./program
+\end{verbatim}
+
+In this example, the trace will be available at path
+{\tt /tmp/some/prefix.pid.eventlog}.
+
+\subsection{ss:instr-runtime-pause}{Pausing and resuming tracing}
+Primitives are available to control events collection at runtime.
+
+{\tt OCAML\_EVENTLOG\_ENABLED} can be set to the {\tt p} flag in order
+to {\em pause} events collection after the program has started.
+
+\begin{verbatim}
+        OCAML_EVENTLOG_ENABLED=p ./program
+\end{verbatim}
+
+The program will have to resume events collection explicitly.
+This can be done via {\tt Gc.eventlog_resume} and
+{\tt Gc.eventlog_pause}, available in the {\stdmoduleref{Gc}} module.
+
+\subsection{ss:instr-runtime-limitations}{Limitations}
+
+The instrumented runtime does not support the {\tt fork} system call.
+A child process forked from an instrumented program will not be traced.
+
+The instrumented runtime aims to provide insight into the runtime's execution
+while maintaining a low overhead.
+However, the impact on performance may increase based on the pressure put on the
+garbage collector.
+
+For a program with an extended running time where the collection of only a
+small sample of events is required, using the {\em eventlog_resume} and
+{\em eventlog_pause} primitives may help relieve some of the
+tracing induced performance impact.
```

Comment:
So what would typically be the tracing you observed when eventlog is paused?

Reply:
If eventlog is paused during a whole run, the tracefile would contain no traces, merely just the Common Trace Format header.

Reply:
When you debug interactive/long running program it's not always evident to do so. I think it would have been nice to provide a standard way to control the collection *from the outside*, modulo a call to a single function, something like: 
```
Gc.install_eventlog_control : ?signum:int -> unit -> unit
``` 
that would install a signal handler so that you can `killall -SIGUSR1 myprogram` collection.

Also there's no way to programmatically flush ?

Reply:
Thank you for your input,

There's no way to programmatically flush, besides relying on the fact that `Gc.eventlog_pause` do force a flush. I think this could be useful indeed.
I think interfacing eventlog controls through signals could be interesting (and signals are indeed the preferred interface for such things), and we could consider it for the next iteration on the instrumented runtime.
(although I think the design may be a bit less straightforward with tracing domains in multicore, ie. is the sighandler installed for every domain, or on only the one executing the code? Can this interface be kept consistent between now and future releases?).

The instrumented runtime having been merged already, I am sorry to have enabled such discussions so late in the process by making the documentation after the fact.
I do not know if additions at this stage are possible (or maybe they could be considered minor tweaks.), however all of this is noted and can be addressed in future iterations of the instrumented runtime.
(as domains will introduce required changes to the instrumentation facility, and provide a complete baseline, feature wise, on which it will be easier to reason in terms of expose-able features that will not need to be revisited later.)

Reply:
My understanding is that we don't really know yet how signals will work in Multicore. I would recommend never having only signals as a way to change things; there should be a programmatic control to use instead of signals for each feature. (One could also think of exporting a control socket, possibly decided through an environment variable, but maybe we can wait for the user need to arise before implementing this.)

> The instrumented runtime having been merged already, I am sorry to have enabled such discussions so late in the process by making the documentation after the fact.

As a strong believer in documentation, I support this message :-)

> I do not know if additions at this stage are possible (or maybe they could be considered minor tweaks.)

Yes, given that this instrumented runtime is completely new, there are no users whose workflow would break yet, so I think you should feel free to propose further changes for now, until the Release Candidate process has started (at this point we want to minimize changes to the release branch).

Reply:
(cc our release gods, @Octachron @damiendoligez)

Reply:
> however all of this is noted and can be addressed in future iterations of the instrumented runtime.

Sure these comments were not meant to be addressed right now. API wise the only real thing I feel is missing is a way to programmatically flush (e.g. because you know appropriate points in your program where it's convenient to do that).

The signal idea (which may indeed not be a good one in the context of multicore) was just me trying to figure out a quick and dirty generic way to be able control instrumentation from the outside.